enhancements/network/openstack-dual-stack.md

```diff
@@ -0,0 +1,175 @@
+
+## Motivation
+
+Customers need dual-stack support to allow migration of workloads from IPv4 to IPv6
+as they are running out of IPv4 addresses. This will allow openshift on openstack to
+serve a wider variety of workloads.
+
+### User Stories
+
+* As an user of a IPI OpenShift cluster, I want to have ingress and egress working with dual-stack.
+* As an user of a IPI OpenShift cluster, I want clusterIP services to be working with dual-stack.
+
+### Goals
+
+- OpenStack VMs are configured with one dual-stack interface.
+- The IPv6 Subnets can be configured with: Stateless, SLAAC or Stateful.
+- The addresses on the VM's interface are configured with the same address
+assigned by OpenStack to the Port.
+- Nodes are configure with both IPv4 and IPv6 addresses, without exposing
+additional addresses, like storage interfaces.
+- Pods are configured with both IPv4 and IPv6 addresses.
+
+### Non-Goals
+
+- Support for OpenStack endpoints configured with dual-stack.
+- Support for LoadBalancer type Services.
+- Installer (IPI) handles creation of the Network and dual-stack Subnet.
+
+## Proposal
+
+Customers will be able to specify the IPv4 and IPv6 Subnets
+meant to be used in the `install-config.yaml`.
+
+### Workflow Description
+
+1. The customers will configure a network (tenant or provider) with IPv4 and
+IPv6 subnets. The IPv6 address mode for the Subnet can be of any type:
+SLAAC, Stateless or Stateful.
+2. The customer will create API and Ingress Ports using the Network created in the
+previous step.
+3. The customer will write the `install-config.yaml` and add the subnet IDs to
+the new field `machinesSubnets`.
+4. The customer will generate the manifests and configured MachineConfigs
+for the master and workers depending on the IPv6 addresses mode chosen for the Subnets
+in step 1. If SLAAC or Stateless was used the MachineConfigs has to contain `ip_addr_mode=eui64`
+and `method=auto` in the connection for the ethernet interface. While when using stateful
+it has to enable `ip=dhcp,dhcpv6` in the kernel arguments.
+5. The `uuid` from the network has to be added in the worker MachineSet alongside with the
+subnets that were pre-filled for the machine-api-provider-openstack to create one unique Port
+with both addresses configured for the Server.
+6. The CIDRs of the subnets have to added to the `cloud-provider-config.yaml`
+in the `address-sort-order` field to make sure IPv4 address will have higher priority in the
+Node addresses.
+7. The installer will validate the subnets and VIPs, add the security group to the api and ingress
+ports and link the floating IPs to those Ports. Also, create the additional IPv6 security group rules
+and configured the servers with dual-stack addresses and add the dual-stack allowed address pairs.
+8. In machine-config-operator, kubelet will be update to not be configured with node-ips.
+9. cloud-provider-openStack will filter the node addresses that are meant to be specified in the Status of the Node
```

Comment:
It's not clear to me who sets the tag. The user? Can we automate this?

Reply:
I'm aware of the context here, but this requires a whole lot of expansion if we're going to include it in this enhancement. I'm happy to work with you on this if it would be helpful.